channeldb/payments_test.go

```diff
@@ -173,3 +174,178 @@ func TestRouteSerialization(t *testing.T) {
+				NumMaxPayments:    2,
+				Reversed:          false,
+				IncludeIncomplete: true,
+			},
+			firstIndex: 1,
+			lastIndex:  2,
+		},
+		// Test query in normal order, start at end, overflow.
+		{
+			query: PaymentsQuery{
+				IndexOffset:       6,
+				NumMaxPayments:    2,
+				Reversed:          false,
+				IncludeIncomplete: true,
+			},
+			firstIndex: 7,
+			lastIndex:  7,
+		},
+		// Start at offset index outside of payments.
+		{
+			query: PaymentsQuery{
+				IndexOffset:       20,
+				NumMaxPayments:    2,
+				Reversed:          false,
+				IncludeIncomplete: true,
+			},
+			firstIndex: 0,
+			lastIndex:  0,
+		},
+		// Test overflow.
+		{
+			query: PaymentsQuery{
+				IndexOffset:       4,
+				NumMaxPayments:    math.MaxUint64,
+				Reversed:          false,
+				IncludeIncomplete: true,
+			},
+			firstIndex: 5,
+			lastIndex:  7,
+		},
+		// Start at offset index outside of payments, reversed order.
+		{
+			query: PaymentsQuery{
+				IndexOffset:       9,
+				NumMaxPayments:    2,
+				Reversed:          true,
+				IncludeIncomplete: true,
+			},
+			firstIndex: 6,
+			lastIndex:  7,
+		},
+		// Test query in reverse order, start at end.
+		{
+			query: PaymentsQuery{
+				IndexOffset:       0,
+				NumMaxPayments:    2,
+				Reversed:          true,
+				IncludeIncomplete: true,
+			},
+			firstIndex: 6,
+			lastIndex:  7,
+		},
+		// Test query in reverse order, starting in middle.
+		{
+			query: PaymentsQuery{
+				IndexOffset:       4,
+				NumMaxPayments:    2,
+				Reversed:          true,
+				IncludeIncomplete: true,
+			},
+			firstIndex: 2,
+			lastIndex:  3,
+		},
+		// Test query in reverse order, starting in middle, with underflow.
+		{
+			query: PaymentsQuery{
+				IndexOffset:       4,
+				NumMaxPayments:    5,
+				Reversed:          true,
+				IncludeIncomplete: true,
+			},
+			firstIndex: 1,
+			lastIndex:  3,
+		},
+		// Test underflow.
+		{
+			query: PaymentsQuery{
+				IndexOffset:       4,
+				NumMaxPayments:    math.MaxUint64,
+				Reversed:          true,
+				IncludeIncomplete: true,
+			},
+			firstIndex: 1,
+			lastIndex:  3,
+		},
+	}
+
+	// Test queried payments.
+	for _, d := range data {
+		querySlice, _ := db.QueryPayments(d.query)
+		if d.firstIndex != querySlice.FirstIndexOffset ||
+			d.lastIndex != querySlice.LastIndexOffset {
+			t.Fatalf("First or last index does not match "+
+				"expected index. Want (%d, %d), got (%d, %d).",
+				d.firstIndex, d.lastIndex,
+				querySlice.FirstIndexOffset,
+				querySlice.LastIndexOffset)
+		}
+	}
+
+	// Fetch all payments and test if order is preserved.
+	querySlice, _ := db.QueryPayments(PaymentsQuery{
+		IndexOffset:       0,
+		NumMaxPayments:    7,
+		Reversed:          true,
+		IncludeIncomplete: true,
+	})
+	for i, p := range allPayments {
+		q := querySlice.Payments[i]
+		if p.sequenceNum != q.sequenceNum {
+			t.Fatalf("sequence numbers do not match, "+
+				"got %v, want %v", p.sequenceNum, q.sequenceNum)
+		}
+	}
+
```

Comment:
May need more test cases handling e.g. different starting `IndexOffset` values.

A naive test to currently trigger an index-out-of-range, see comment above:

```golang
// Should return an empty slice but currently triggers an index-out-of-range.
querySlice, _ = db.QueryPayments(PaymentsQuery{
	IndexOffset:       7,
	NumMaxPayments:    7,
	Reversed:          false,
	IncludeIncomplete: true,
})
if len(querySlice.Payments) != 0 {
	t.Fatalf("index-out-of-range is triggered before test is completed.")
}
```

Reply:
Perfect, test triggered the out-of-range error. Added the test, which is now passing.

Reply:
Hey @bjarnemagnussen, thank you very much for your review!